### PR TITLE
Adding JSONHelpers tests and SemanticVersion Tests

### PR DESCRIPTION
**Pull request by YangLyu89**
re-factor the CompletionController and got some tests for it as well as the SemanticVersion

### CODE REVIEW

File: test/LibraryManager.Test/Microsoft.Web.LibraryManager.Test.csproj [hunk 1]
```diff
@@ -17,6 +17,14 @@
...
         <ProjectReference Include="..\LibraryManager.Mocks\Microsoft.Web.LibraryManager.Mocks.csproj" />
     </ItemGroup>
+    <ItemGroup>
+      <Reference Include="Microsoft.CSS.Core">
+        <HintPath>..\..\..\..\LibraryManager\lib\Microsoft.CSS.Core.dll</HintPath>
+      </Reference>
+      <Reference Include="Microsoft.JSON.Core">
+        <HintPath>..\..\lib\Microsoft.JSON.Core.dll</HintPath>
+      </Reference>
+    </ItemGroup>
```
Comment (bhsubra) [inline test/LibraryManager.Test/Microsoft.Web.LibraryManager.Test.csproj]: As discussed offline, these are very old. We made some api's internal in JSON and CSS since then. Using this is very risky.

**Comment by bhsubra** [comment]
>     </Reference>

Not related to your change, but wanted to make a note so we can discuss in group code review meeting. These are very old. 

---
Refers to: src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj:181 in e084c0d. [](commit_id = e084c0dadb2c764242431b2d07fac3750be57215, deletion_comment = False)

File: src/LibraryManager/Microsoft.Web.LibraryManager.csproj [hunk 2]
```diff
@@ -13,6 +13,12 @@
     <PackageReference Include="Newtonsoft.Json" />
     <PackageReference Include="System.ValueTuple" Condition="'$(TargetFramework)' != 'netstandard2.0'" />
     <ProjectReference Include="..\LibraryManager.Contracts\Microsoft.Web.LibraryManager.Contracts.csproj" />
+    <Reference Include="Microsoft.CSS.Core">
+      <HintPath>..\..\..\..\LibraryManager\lib\Microsoft.CSS.Core.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.JSON.Core">
+      <HintPath>..\..\lib\Microsoft.JSON.Core.dll</HintPath>
+    </Reference>
```
Comment (bhsubra) [inline src/LibraryManager/Microsoft.Web.LibraryManager.csproj]: Same as other comment

File: src/LibraryManager/Helpers/JsonHelpers.cs [hunk 3]
```diff
@@ -1,15 +1,69 @@
...
+using Microsoft.JSON.Core.Parser;
+using Microsoft.JSON.Core.Parser.TreeItems;
+using Microsoft.Web.LibraryManager.Contracts;
 
-namespace Microsoft.Web.LibraryManager.Vsix
+namespace Microsoft.Web.LibraryManager.Helpers
 {
     internal static class JsonHelpers
     {
+        internal static JSONParseItem GetItemBeforePosition(int pos, JSONComplexItem parentItem)
```
Comment (prafullbhosale) [inline src/LibraryManager/Helpers/JsonHelpers.cs]: >JSONParseItem [](start = 24, length = 13)

The moved files should remain in the VSIX project. 
For adding the tests, you can add new Test project which refers the VSIX project.

File: src/LibraryManager/Microsoft.Web.LibraryManager.csproj [hunk 4]
```diff
@@ -13,6 +13,15 @@
...
     <ProjectReference Include="..\LibraryManager.Contracts\Microsoft.Web.LibraryManager.Contracts.csproj" />
+    <Reference Include="Microsoft.CSS.Core, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\lib\Microsoft.CSS.Core.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Microsoft.JSON.Core">
+      <HintPath>..\..\lib\Microsoft.JSON.Core.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
```
Comment (prafullbhosale) [inline src/LibraryManager/Microsoft.Web.LibraryManager.csproj]: Revert this. 
This project shouldn't take any dependencies on VS assemblies

File: src/LibraryManager/Helpers/SemanticVersion.cs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ﻿using System;
 
-namespace Microsoft.Web.LibraryManager.Providers.Unpkg
+namespace Microsoft.Web.LibraryManager.Helpers
```
Comment (prafullbhosale) [inline src/LibraryManager/Helpers/SemanticVersion.cs]: >Helpers [](start = 39, length = 7)

Move this to its own folder named something like 'Versioning'

File: src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj [hunk 6]
```diff
@@ -100,7 +100,6 @@
     <Compile Include="Json\Completion\PathCompletionProvider.cs" />
     <Compile Include="Json\Completion\ProviderCompletionProvider.cs" />
     <Compile Include="Json\Completion\CompletionController.cs" />
-    <Compile Include="Json\JsonHelpers.cs" />
```
Comment (prafullbhosale) [inline src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj]: Add this one back.

File: src/LibraryManager/Helpers/JsonHelpers.cs [hunk 7]
```diff
@@ -1,15 +1,69 @@
...
+                        // Recurse to find the deepest item
+                        item = GetItemBeforePosition(pos, complexItem);
+                    }
+                }
+            }
+
+            return item;
+        }
+
+        internal static int FindInsertIndex(JSONParseItemList jsonParseItems, int rangeStart, bool beforeExisting)
```
Comment (prafullbhosale) [inline src/LibraryManager/Helpers/JsonHelpers.cs]: If this always going to be passed in as true, this flag can be removed.

File: src/LibraryManager/Helpers/JsonHelpers.cs (hunk 3 above)
Comment (YangLyu89) [reply src/LibraryManager/Helpers/JsonHelpers.cs]: Yes. will do

---
In reply to: [203547823](https://github.com/aspnet/LibraryManager/pull/324#discussion_r203547823) [](ancestors = 203547823)

File: src/LibraryManager/Helpers/SemanticVersion.cs (hunk 5 above)
Comment (YangLyu89) [reply src/LibraryManager/Helpers/SemanticVersion.cs]: Create SemanticVersion Folder

---
In reply to: [203548469](https://github.com/aspnet/LibraryManager/pull/324#discussion_r203548469) [](ancestors = 203548469)

File: test/LibraryManager.Test/Helpers/JSONHelpersTest.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        [DataRow(1, "{")]
+        [DataRow(171, "}")]
+        public void JSONHelpers_GetItemBeforePosition(int position, string expextedText)
+        {
+            JSONComplexItem complexItem = JSONParser.Parse(_jsonText);
+
+            JSONParseItem item = JsonHelpers.GetItemBeforePosition(position, complexItem);
+
+            Assert.IsTrue(item is JSONTokenItem);
+            Assert.AreEqual(expextedText, item.Text);
```
[the remark this answers is not in the document]
Comment (YangLyu89) [reply test/LibraryManager.Test/Helpers/JSONHelpersTest.cs]: will change.

---
In reply to: [203549044](https://github.com/aspnet/LibraryManager/pull/324#discussion_r203549044) [](ancestors = 203549044)

File: src/LibraryManager/Helpers/JsonHelpers.cs (hunk 7 above)
Comment (YangLyu89) [reply src/LibraryManager/Helpers/JsonHelpers.cs]: I think it can be removed.

---
In reply to: [203551539](https://github.com/aspnet/LibraryManager/pull/324#discussion_r203551539) [](ancestors = 203551539)

File: src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj [hunk 9]
```diff
@@ -53,6 +53,7 @@
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="Commands\InstallLibraryCommand.cs" />
+    <Compile Include="Json\JsonHelpers.cs" />
```
Comment (YangLyu89) [inline src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj]: undo the change.

**Comment by YangLyu89** [comment]
> using System.Linq;

Sorting the namespace

---
Refers to: src/LibraryManager.Vsix/Json/Completion/BaseCompletionProvider.cs:14 in a702b43. [](commit_id = a702b430e42b0fe96d152eee71ec4e9dd61c6760, deletion_comment = False)
